// speech-to-text/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var SpeechToText =
/*#__PURE__*/
function () {
  /*
    Arguments for the constructor:
     - onFinalised - a callback that will be passed the finalised transcription from the cloud. Slow, but accuate.
    - onAnythingSaid - a callback that will be passed interim transcriptions. Fairly immediate, but less accurate than finalised text.
    - language - the language to interpret against. Default is US English.
     */
  function SpeechToText(onFinalised, onAnythingSaid) {
    var _this = this;

    var language = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'en-US';

    _classCallCheck(this, SpeechToText);

    // Check to see if this browser supports speech recognition
    // https://developer.mozilla.org/en-US/docs/Web/API/SpeechRecognition#Browser_compatibility
    if (!('webkitSpeechRecognition' in window)) {
      throw new Error("This browser doesn't support speech recognition. Try Google Chrome.");
    }

    var continuouslyListen = !!onAnythingSaid;
    var SpeechRecognition = window.webkitSpeechRecognition;
    this.recognition = new SpeechRecognition(); //  if to continuously listen, get interim results too

    this.recognition.continuous = continuouslyListen;
    this.recognition.interimResults = continuouslyListen;
    this.recognition.lang = language;
    var finalTranscript = ''; // process both interim and finalised results

    this.recognition.onresult = function (event) {
      var interimTranscript = ''; // concatenate all the transcribed pieces together (SpeechRecognitionResult)

      for (var i = event.resultIndex; i < event.results.length; i += 1) {
        var transcriptionPiece = event.results[i][0].transcript; // check for a finalised transciption in the cloud

        if (event.results[i].isFinal) {
          finalTranscript += transcriptionPiece;
          onFinalised(finalTranscript);
          finalTranscript = '';
        } else if (continuouslyListen) {
          interimTranscript += transcriptionPiece;
          onAnythingSaid(interimTranscript);
        }
      }
    }; // if speech recognition ends, and it's flagged to restart, then restart!


    this.recognition.onend = function () {
      if (_this.shouldRestart) {
        _this.startListening();

        _this.shouldRestart = false;
      }
    };

    this.recognition.onerror = function (evt) {
      console.log("Error with speech recognition: ".concat(evt.error, ". Restarting service..."));

      _this.restart();
    };
  }

  _createClass(SpeechToText, [{
    key: "startListening",
    value: function startListening() {
      this.recognition.start();
    }
  }, {
    key: "stopListening",
    value: function stopListening() {
      this.recognition.stop();
    } // to restart the speech recognition engine, stop it, then flag that it should restart.

  }, {
    key: "restart",
    value: function restart() {
      this.stopListening();
      this.shouldRestart = true;
    }
  }]);

  return SpeechToText;
}();

exports.default = SpeechToText;
